attachment_preview/model/ir_attachment.py

```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    OpenERP, Open Source Management Solution
+#    This module copyright (C) 2014 Therp BV (<http://therp.nl>).
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+import collections
+import os.path
+import mimetypes
+import base64
+from openerp.osv.orm import Model
+
+
+class IrAttachment(Model):
+    _inherit = 'ir.attachment'
+
+    def get_binary_extension(
+            self, cr, uid, model, ids, binary_field, filename_field=None,
+            context=None):
+        result = {}
+        for this in self.pool[model].browse(
+                cr, uid,
+                ids if isinstance(ids, collections.Iterable) else [ids],
+                context=context):
+            if not this.id:
+                result[this.id] = False
+                continue
+            extension = ''
+            if filename_field and this[filename_field]:
+                filename, extension = os.path.splitext(this[filename_field])
+            if not this[binary_field]:
+                result[this.id] = False
+                continue
+            if not extension:
+                try:
+                    import magic
+                    ms = magic.open(
+                        hasattr(magic, 'MAGIC_MIME_TYPE')
+                        and magic.MAGIC_MIME_TYPE or magic.MAGIC_MIME)
+                    ms.load()
+                    mimetype = ms.buffer(
+                        base64.b64decode(this[binary_field]))
+                except ImportError:
+                    (mimetype, encoding) = mimetypes.guess_type(
+                        'data:;base64,' + this[binary_field], strict=False)
```

Comment:
Is this method of guessing the mime type still optimal ?

We have now the method [guess_mimetype](https://github.com/odoo/odoo/blob/12.0/odoo/tools/mimetypes.py#L135) in tools.

It fails when `python-magic` is not installed. 

In the section after `ImportError`, the string literal: `'data:;base64,'` is now _unicode_ in **python3** and fails with one **unicode** error instead of trying another guess as intended by the code.

Reply:
I mean The code is supposed to have an alternate method to that `magic`. That code is not being included in the coverage and it is failing.

Reply:
Tested in runbot. Works very well.
Any plans to let the split screen functionality being auto-triggered by the 
`<div class="o_attachment_preview" attrs="{'invisible': ['|',('type', '!=', 'in_invoice'),('state', '!=', 'draft')]}"/>`
 found in Odoo Core in various places (e.g. on Vendor bills)?

Reply:
@hhgabelgaard:   I guess you wanted to make a functional review and ended commented my review.

I am not talking about the functional part of the module here. I bet it is a good work done there. What I am pointing out here is not something that you can see in the user interface. it is about the way to handle dependencies.

Runbot works ok, because the new [code](https://github.com/OCA/knowledge/blob/063c58e25a46589f39d972fea6350328e4f40a5f/.travis.yml#L33) tweaks the **runbot** installation process to install `python-magic`.

When python-magic is not installed the code fails. And python magic seems to be marked as a dependency of Odoo in [v12](https://github.com/odoo/odoo/blob/12.0/requirements.txt), nor from the **addon** dependencies.

Reply:
No answer for this ? it is still giving error in our tests instances.